

library(dplyr)
library(ggplot2)
library(infercnv)
library(phytools)
library(dendextend)
library(Seurat)


geco.load <- function (filename) # File to be loaded
{
  if (file.exists(filename)) 
    return(eval(parse(text = load(filename))))
  cat(paste("error - function geco.load : file ", filename, 
            " not found"))
  NULL
}

factoall <- function (d, ncmax = 10) 
{
  n <- ncol(d)
  for (i in 1:n) {
    if (is.factor(d[, i])) {
      d[, i] <- as.character(d[, i])
      na <- which(is.na(d[, i]))
      num <- suppressWarnings(as.numeric(d[, i]))
      nanum <- which(is.na(num))
      if (length(nanum) == length(na)) {
        #                int <- suppressWarnings(as.integer(d[, i]))
        #                naint <- which(is.na(int))
        #                nc <- nchar(num)
        #                if (length(naint) == length(nanum) & all(nc < ncmax)) {
        #                  d[, i] <- int
        #                }
        #                else {
        d[, i] <- num
        #                }
      }
    }
  }
  d
}

scale_fill_custom <- function(...){
  ggplot2:::manual_scale(
    'fill', 
    values = setNames(c("red", "orange", "yellow", "grey20", "grey"), c("mutation, >=3 SNV reads", "mutation, 2 SNV reads", "mutation, 1 SNV read", "no mutation", "not known")), 
    ...
  )
}


######################
##### PARAMETERS #####
######################

name_sample <- c("CHC2959N", "CHC2959T", "CHC2960T", "CHC3133T", "CHC3377N", "CHC3377T", "CHC3610T", "CHC3662T")

datadir <- "F:/Amelie-Datas/Pediatric tumors/3.Multiome ATAC RNA/2.snRNAseq/9.Clonal_evolution/11p15/0.Input_data/"
resdir <- file.path("F:/Amelie-Datas/Pediatric tumors/3.Multiome ATAC RNA/2.snRNAseq/9.Clonal_evolution/11p15/")
seurat_dir_all <- file.path("F:/Amelie-Datas/Pediatric tumors/3.Multiome ATAC RNA/2.snRNAseq/6.Clean_cluster_definition_tumor_immune/")


###########################
##### 0. DATA LOADING #####
###########################

### snMultiome info ###
recap_by_cell <- geco.load(file.path(datadir, "recap_by_cell_all_samples.Rdata")) # table generated by Théo with 11p15 info for cells (tumor+non tumor) of all samples, on unfiltered BAM (no filter on xf=25) 

# Add sample name to cell ID
for(n in names(recap_by_cell)){
  recap_by_cell[[n]] <- as.data.frame(recap_by_cell[[n]]) # convert tibble to dataframe
  recap_by_cell[[n]][,"ReadGroup"] <- paste(n, recap_by_cell[[n]][,"ReadGroup"], sep="_")
  rownames(recap_by_cell[[n]]) <- recap_by_cell[[n]][,"ReadGroup"]
}
  
### Load Seurat objects ###
sample_seurat_all <- geco.load(file.path(seurat_dir_all, "Seurat_object_analyses_all_cells.RData")) # all cells


###############################################
##### 1. VISUALIZE CNLOH ON COMPLETE UMAP #####
###############################################

##### 1. Add cnLOH info in seurat object #####
#---------------------------------------------

sample_seurat_all$cnLOH <- "unknown"
for(n in names(recap_by_cell)){
  final_cells <- intersect(colnames(sample_seurat_all), recap_by_cell[[n]][,"ReadGroup"]) # restrict to cells post-QC (screadcounts gives info for whole BAM = no QC on cells)
  sample_seurat_all@meta.data[final_cells,"cnLOH"] <- recap_by_cell[[n]][final_cells,"status_cnLOH"]
}


##### 2. Plot #####
#------------------
df_UMAP <- as.data.frame(sample_seurat_all@reductions$umap@cell.embeddings[,c("UMAP_1", "UMAP_2")]) # extract the info to plot
if(identical(rownames(df_UMAP), rownames(sample_seurat_all@meta.data))){df_UMAP$cnLOH <- sample_seurat_all$cnLOH} # add cnLOH info
df_UMAP <- df_UMAP[colnames(subset(sample_seurat_all, subset=orig.ident%in%name_sample & selection_proposition_strict != "undefined")),] # remove undefined cells (doublets etc) for plot
df_UMAP$cnLOH <- factor(df_UMAP$cnLOH, levels=c("cnLOH high confidence", "cnLOH low confidence", "no cnLOH high confidence",  "no cnLOH low confidence", "unknown")) # order for plot
df_UMAP <- df_UMAP[sample(rownames(df_UMAP)),] # sample the cell names for the plot

pdf(file.path(resdir, "UMAP_all_cells_11p15_cnLOH.pdf"))
print(DimPlot(subset(sample_seurat_all, subset=orig.ident%in%name_sample & selection_proposition_strict != "undefined"), group.by = "orig.ident"))
print(ggplot(df_UMAP, aes(x=UMAP_1, y=UMAP_2, col=cnLOH))+
  geom_point(size=1.3)+
  theme_classic()+
  scale_color_manual(values=c("red", "red", "forestgreen", "forestgreen", "grey95"))+
  theme(axis.text=element_text(size=12), axis.title=element_text(size=14), legend.position = "none"))
print(DimPlot(subset(sample_seurat_all, subset=orig.ident%in%name_sample & selection_proposition_strict != "undefined"), group.by = "orig.ident"))
ggplot(df_UMAP, aes(x=UMAP_1, y=UMAP_2, col=cnLOH))+
  geom_point(size=1.3)+
  theme_classic()+
  scale_color_manual(values=c("red", "red", "forestgreen", "forestgreen", "grey95"))+
  theme(axis.text=element_text(size=12), axis.title=element_text(size=14))
dev.off()


